core/src/Kokkos_MemoryPool.hpp

```diff
@@ -156,8 +156,13 @@ class MemoryPool {
     size_t reserved_bytes;   ///<  Unallocated bytes in assigned superblocks
   };
 
+  // This function is templated to avoid needing a full definition of
+  // DefaultHostExecutionSpace at class instantiation
+  template <typename ExecutionSpace = Kokkos::DefaultHostExecutionSpace>
```

Comment:
Can't we move the definition outside of the class and place it after DefaultHostExecutionSpace is a complete type?

Reply:
The class is templated so we can't put it in a source file without explicit template instantiations. I'm not sure that creating a separate header file would be preferrable over the solution here.

Reply:
I just don't like templating the function and then static_asserting on the type. I don't see any problem with having an impl header

Reply:
Note that we are already doing this in https://github.com/kokkos/kokkos/pull/7519 (which caused the issue) in other places. I just don't want to figure out the appropriate include order to make all of this work. Feel free to give it a shot if you have a better solution!